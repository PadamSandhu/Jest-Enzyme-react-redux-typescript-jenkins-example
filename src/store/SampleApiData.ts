export const mockResponse = {
    time: {
        updated: "Feb 21, 2018 20:56:00 UTC",
        updatedISO: "2018-02-21T20:56:00+00:00",
        updateduk: "Feb 21, 2018 at 20:56 GMT"
    },
    disclaimer: "This data was produced from the CoinDesk Bitcoin Price Index (USD). Non-USD currency data converted using hourly conversion rate from openexchangerates.org",
    bpi: {
        USD: {
            code: "USD",
            rate: "10,329.3075",
            description: "United States Dollar",
            rate_float: 10329.3075
        },
        CAD: {
            code: "CAD",
            rate: "13,101.5800",
            description: "Canadian Dollar",
            rate_float: 13101.58
        }
    }
};