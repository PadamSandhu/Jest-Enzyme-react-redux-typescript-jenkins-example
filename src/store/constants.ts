export const SET_BALANCE = "SET_BALANCE";
export const DEPOSIT = "DEPOSIT";
export const WITHDRAW = "WITHDRAW";
export const BIT_COIN_API_CALL = "BIT_COIN_API_CALL";