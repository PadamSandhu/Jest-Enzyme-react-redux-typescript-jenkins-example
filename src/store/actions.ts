import * as constants from "./constants";

export const setBalance = (balance) => ({
  type: constants.SET_BALANCE,
   payload: balance
});

export const deposit = (deposit) => ({
  type: constants.DEPOSIT,
  payload: deposit
});

export const withdraw = (withdraw) => ({
  type: constants.WITHDRAW,
  payload: withdraw
});

export const bitcoinapicall = (balance) => ({
  type: constants.BIT_COIN_API_CALL,
  payload: balance
});
