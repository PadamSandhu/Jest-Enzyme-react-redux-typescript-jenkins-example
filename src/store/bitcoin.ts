import { BIT_COIN_API_CALL } from "./constants";

export const fetchBitcoin = () => {
    return dispatch => {
        return fetch("https://api.coindesk.com/v1/bpi/currentprice/CAD.json")
        .then(response => response.json())
        .then(json => dispatch({ type: BIT_COIN_API_CALL, payload: json }));
    };
};