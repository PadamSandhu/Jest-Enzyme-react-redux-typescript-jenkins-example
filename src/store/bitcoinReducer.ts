import { BIT_COIN_API_CALL } from "./constants";

const initialState = {

};

export default (state = initialState, action) => {
  switch (action.type) {

  case BIT_COIN_API_CALL:
    return action.payload;

  default:
    return state;
  }
};
