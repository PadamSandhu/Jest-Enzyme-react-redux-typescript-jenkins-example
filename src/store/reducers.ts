import * as constants from "./constants";
import * as Cookies from "js-cookie";

const BALANCE_COOKIE = "BALANCE_COOKIE";  // Uniqure Id string for this cookie !!
// Cookies allow us to store information on the userâ€™s browser with a unique string.

export default (state = 0, action) => {
  let balance;

  switch (action.type) {

  case constants.SET_BALANCE:
    balance =  action.payload;
    break;
  case constants.DEPOSIT:
    balance = (action.payload + state);
    break;
    case constants.WITHDRAW:
    balance = ( state - action.payload);
    break;
  default:
    balance = parseInt(Cookies.get(BALANCE_COOKIE), 10) || state;
  }
  Cookies.set(BALANCE_COOKIE, balance);   // With any change in the state we are updating our cookie !!

  return balance;
};
