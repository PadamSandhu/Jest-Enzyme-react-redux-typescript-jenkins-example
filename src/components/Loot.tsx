import * as React from "react";
import { connect } from "react-redux";
import {fetchBitcoin} from "../store/bitcoin";

export class Loot extends React.Component<any, any> {
    componentDidMount() {
        this.props.fetchBitcoin();
    }
     computeBitcoin = () => {
        const { bitcoin } = this.props;
        if (Object.keys(bitcoin).length === 0) {
            return ""; }
            else {
                return parseInt(this.props.balance, 10) / parseInt(this.props.bitcoin.bpi.CAD.rate.replace(",", ""), 10);
            } /// This is a good source of error !!
    }

    render() {
        return (
            <div>
                <h3 className="bitcoinBalance" >Bitcoin balance: {this.computeBitcoin()}</h3>
            </div>
        );
    }
}

export default connect( state => state , { fetchBitcoin } )(Loot);  /// We Don't need to test this function because we already tested
