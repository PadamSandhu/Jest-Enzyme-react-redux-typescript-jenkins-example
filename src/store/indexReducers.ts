import { combineReducers } from "redux";
import balance from "./reducers";
import bitcoin from "./bitcoinReducer";

export default combineReducers({ balance, bitcoin });