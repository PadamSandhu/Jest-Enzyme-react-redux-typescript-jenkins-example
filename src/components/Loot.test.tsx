import * as React from "react";
import { shallow } from "enzyme";
import { Loot } from "./Loot";
import { mockResponse } from "../store/SampleApiData";

describe("Loot", () => {
    const mockFetchbitcoin = jest.fn();

    let props: {balance: number, bitcoin: any, fetchBitcoin: any } = { balance: 10, bitcoin: {}, fetchBitcoin: mockFetchbitcoin };
    let loot = shallow(<Loot {...props}/>);

    it("Renders Properly", () => {
        expect(loot).toMatchSnapshot();
    });

    // Writing a test if this component will mount !!


    describe("when mounted", () => {

        it("dispathces the `fetchBitcoin()` method it receives from props", () => {
            expect(mockFetchbitcoin).toHaveBeenCalled();
        });
    });

    describe("When there are valid bitcoin props" , () => {
        beforeEach(() => {
            props = { balance: 14000, bitcoin: mockResponse, fetchBitcoin: mockFetchbitcoin};
            loot = shallow(<Loot {...props}/>);
        });
        it("Displays the correct bitcoin", () => {
            const BitCoinConversion = (props.balance) / parseInt(props.bitcoin.bpi.CAD.rate.replace(",", ""), 10);
            expect(loot.find(".bitcoinBalance").text()).toEqual("Bitcoin balance: " + BitCoinConversion);
        });
    });
});