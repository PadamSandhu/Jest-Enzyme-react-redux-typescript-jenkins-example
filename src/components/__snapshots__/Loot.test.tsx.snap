// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Loot Renders Properly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Loot
    balance={10}
    bitcoin={Object {}}
    fetchBitcoin={
      [MockFunction] {
        "calls": Array [
          Array [],
        ],
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <h3
        className="bitcoinBalance"
      >
        Bitcoin balance: 
        
      </h3>,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          "Bitcoin balance: ",
          "",
        ],
        "className": "bitcoinBalance",
      },
      "ref": null,
      "rendered": Array [
        "Bitcoin balance: ",
        "",
      ],
      "type": "h3",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <h3
          className="bitcoinBalance"
        >
          Bitcoin balance: 
          
        </h3>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "Bitcoin balance: ",
            "",
          ],
          "className": "bitcoinBalance",
        },
        "ref": null,
        "rendered": Array [
          "Bitcoin balance: ",
          "",
        ],
        "type": "h3",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
